### ISSUE TITLE
Segmentation fault with mingw-w64-x86_64-openblas package

### ISSUE THREAD

**Issue by machakann** [msys2/MINGW-packages#12857]
### Description

I encountered a segmentation fault when I developed a c++ program using LAPACK zheev driver included in the openblas package (`-lopenblas` option). I'm not an expert of c++ but at least the same program ran correctly with LAPACK with reference BLAS (`-llapack -lblas` options with mingw-w64-x86_64-lapack package).

### Verification

- [X] I have verified that my MSYS2 is up-to-date before submitting the report (see https://www.msys2.org/docs/updating/)

### Windows Version

Microsoft Windows [Version 10.0.22000.856]

### MINGW environments affected

- [X] MINGW64
- [ ] MINGW32
- [ ] UCRT64
- [ ] CLANG64
- [ ] CLANG32
- [ ] CLANGARM64

### Expected behavior

The compiled program runs without error and finishes silently.

### Actual behavior

The compiled program echoes "Segmentation fault" and finished with a non-zero exit code.

### Repro steps

I have used *Windows sandbox* to try the following steps in a clean environment. However, the segmentation fault should be reproduced even without using the sandbox.

1. Install MSYS2 from https://www.msys2.org/
2. Run `pacman -Syu` twice
3. Run `pacman -S mingw-w64-x86_64-gcc mingw-w64-x86_64-openblas`
4. Run the following
    ```
    mkdir test_openblas
    cd test_openblas
    nano test_openblas.cpp
    g++ test_openblas.cpp -o a.out -lopenblas
    ./a.out
    ```

The content of the *test_openblas.cpp* is shown below.
```cpp
#include <complex>
#include <vector>

using complex = std::complex<double>;

extern "C" {
  void zheev_(const char& JOBZ, const char& UPLO, const int& N, complex* A,
              const int& LDA, double* W, complex* WORK, const int& LWORK,
              double* RWORK, int& INFO);
};


int main() {
    using std::vector;
    constexpr int n = 256;
    vector<complex> a(n*n);
    int lda = n;
    double w[n];
    int lwork;
    complex work_temp[1];
    double rwork[3*n - 2];
    int info;
    zheev_('V', 'U', n, a.data(), lda, w, work_temp, -1, rwork, info);
    lwork = (int) work_temp[0].real();
    auto work = new complex[lwork];
    zheev_('V', 'U', n, a.data(), lda, w, work, lwork, rwork, info); // Segmentation fault
    return info;
}
```

A strange thing I found is that it runs correctly when `n = 255`.

**Comment by mati865** [comment]
Possibly GCC bug?
 Under MINGW64 environment it fails with:
```
* thread #1, stop reason = Exception 0xc0000005 encountered at address 0x7ffab19ab690: Access violation reading location 0x201e19b7010
    frame #0: 0x00007ffab19ab690 libopenblas.dll`zgemv_n_ZEN + 4992
libopenblas.dll`zgemv_n_ZEN:
->  0x7ffab19ab690 <+4992>: vinsertf128 $0x1, (%r9), %ymm0, %ymm0
    0x7ffab19ab696 <+4998>: movq   0x88(%rsp), %rdi
    0x7ffab19ab69e <+5006>: vpermpd $0x11, %ymm0, %ymm2       ; ymm2 = ymm0[1,0,1,0]
    0x7ffab19ab6a4 <+5012>: vpermpd $0x44, %ymm0, %ymm0       ; ymm0 = ymm0[0,1,0,1]
(lldb) bt
* thread #1, stop reason = Exception 0xc0000005 encountered at address 0x7ffab19ab690: Access violation reading location 0x201e19b7010
  * frame #0: 0x00007ffab19ab690 libopenblas.dll`zgemv_n_ZEN + 4992
    frame #1: 0x00007ffab037bbff libopenblas.dll`zgemv_ + 1039
    frame #2: 0x00007ffab27c65a7 libopenblas.dll`zlatrd_ + 2455
    frame #3: 0x00007ffab272f6fd libopenblas.dll`zhetrd_ + 2093
    frame #4: 0x00007ffab272461b libopenblas.dll`zheev_ + 859
    frame #5: 0x00007ff6ae741690 test_openblas.exe`main + 544
    frame #6: 0x00007ff6ae7413d7 test_openblas.exe`__tmainCRTStartup at crtexe.c:329:15
    frame #7: 0x00007ff6ae741436 test_openblas.exe`mainCRTStartup at crtexe.c:206:9
    frame #8: 0x00007ffbbd1754e0 kernel32.dll`BaseThreadInitThunk + 16
    frame #9: 0x00007ffbbde2485b ntdll.dll`RtlUserThreadStart + 43
```
But under CLANG64 environment where OpenBLAS it runs fine.

I don't know OpenBLAS source code but other option is OpenBLAS/Lapack issue since LAPACK is enabled only in GCC based subsystems.

**Comment by mmuetzel** [comment]
I can also confirm the crash in MINGW64 with the package installed from MSYS2. However, if I rebuild locally from `PKGBUILD`, the crash disappears.
Maybe somehow some features from the build system leak into the library even though we build with `DYNAMIC_ARCH`?

> I don't know OpenBLAS source code but other option is OpenBLAS/Lapack issue since LAPACK is enabled only in GCC based subsystems.

Afaict, LAPACK is built as a part of OpenBLAS since a couple of months in MSYS2 for all environments. The CLANG environments build an f2c translated version of the reference LAPACK implementation though. The GCC environment build the "original" Fortran version.

**Comment by mati865** [comment]
>  Maybe somehow some features from the build system leak into the library even though we build with `DYNAMIC_ARCH`?

I really doubt, as you can see in the backtrace it called `zgemv_n_ZEN` (I have Zen3 CPU) and our CI runs on variety of Xeon CPUs (some of them don't even have AVX2).

> However, if I rebuild locally from PKGBUILD, the crash disappears.

Maybe GCC/GFortran bug? Would be worth rebuilding it on CI and testing whether it fixes crash on other machines.

**Comment by MehdiChinoune** [comment]
> I really doubt, as you can see in the backtrace it called `zgemv_n_ZEN` (I have Zen3 CPU) and our CI runs on variety of Xeon CPUs (some of them don't even have AVX2).

All of them have AVX2. GHA run on Azure Standard_DS2_v2. The oldest of them is  Intel® Xeon® E5-2673 v3 2.4 GHz (Haswell).
https://docs.github.com/en/actions/using-github-hosted-runners/about-github-hosted-runners#cloud-hosts-used-by-github-hosted-runners
https://docs.microsoft.com/en-us/azure/virtual-machines/dv2-dsv2-series#dsv2-series

**Comment by mmuetzel** [comment]
I triggered a rebuild with the CI and installed the MINGW64 build artifact from it: https://github.com/mmuetzel/MINGW-packages/actions/runs/2939366959

That still crashes with a segmentation fault for the test program.

The backtrace looks pretty similar to yours:
```
$ lldb ./a.out
(lldb) target create "./a.out"
Current executable set to 'C:\msys64\home\Markus\a.out' (x86_64).
(lldb) r
Process 26144 launched: 'C:\msys64\home\Markus\a.out' (x86_64)
Process 26144 stopped
* thread #1, stop reason = Exception 0xc0000005 encountered at address 0x7fffe5adf5a3: Access violation reading location 0x20ef1f9e010
    frame #0: 0x00007fffe5adf5a3 libopenblas.dll`zgemv_n_ZEN + 5187
libopenblas.dll`zgemv_n_ZEN:
->  0x7fffe5adf5a3 <+5187>: vinsertf128 $0x1, (%r9), %ymm0, %ymm0
    0x7fffe5adf5a9 <+5193>: movq   0x88(%rsp), %rdi
    0x7fffe5adf5b1 <+5201>: vpermpd $0x11, %ymm0, %ymm2       ; ymm2 = ymm0[1,0,1,0]
    0x7fffe5adf5b7 <+5207>: vpermpd $0x44, %ymm0, %ymm0       ; ymm0 = ymm0[0,1,0,1]
(lldb) bt
error: a.out {0x000022af}: DIE has DW_AT_ranges(DW_FORM_sec_offset 0x1a) attribute, but range extraction failed (invalid range list offset 0x1a), please file a bug and attach the file at the start of this error message
error: a.out {0x000022af}: DIE has DW_AT_ranges(DW_FORM_sec_offset 0x1a) attribute, but range extraction failed (invalid range list offset 0x1a), please file a bug and attach the file at the start of this error message
* thread #1, stop reason = Exception 0xc0000005 encountered at address 0x7fffe5adf5a3: Access violation reading location 0x20ef1f9e010
  * frame #0: 0x00007fffe5adf5a3 libopenblas.dll`zgemv_n_ZEN + 5187
    frame #1: 0x00007fffe44abbff libopenblas.dll`zgemv_ + 1039
    frame #2: 0x00007fffe68fbd67 libopenblas.dll`zlatrd_ + 2455
    frame #3: 0x00007fffe6864ebd libopenblas.dll`zhetrd_ + 2093
    frame #4: 0x00007fffe6859ddb libopenblas.dll`zheev_ + 859
    frame #5: 0x00007ff6007818c5 a.out`main
    frame #6: 0x00007ff6007813ae a.out`__tmainCRTStartup
    frame #7: 0x00007ff6007814e6 a.out`mainCRTStartup
    frame #8: 0x00007ff8795254e0 kernel32.dll`BaseThreadInitThunk + 16
    frame #9: 0x00007ff87b46485b ntdll.dll`RtlUserThreadStart + 43
```

Maybe I'm misunderstanding what you are saying. But IIUC it might still be the case that the common code of the multiarch library is preparing for CPU features that the build system (GitHub's CI) supports but the host system (my PC) doesn't. But that is just a guess admittedly.

**Comment by mmuetzel** [comment]
Maybe an important detail that I didn't write before: In the local rebuild, I didn't strip the binaries.

**Comment by MehdiChinoune** [comment]
> Maybe an important detail that I didn't write before: In the local rebuild, I didn't strip the binaries.

You could disable stripping in PKGBUILD by adding `!strip` to options.

**Comment by mati865** [comment]
> Maybe I'm misunderstanding what you are saying. But IIUC it might still be the case that the common code of the multiarch library is preparing for CPU features that the build system (GitHub's CI) supports but the host system (my PC) doesn't. But that is just a guess admittedly.

There are no instructions available on CI that ZEN doesn't support which would help with performance in this library.
`vinsertf128` seen in this backtrace comes from AVX (not even AXV2) and there is no illegal instruction error.

**Comment by mmuetzel** [comment]
I let the CI rebuild again disabling stripping and installed the MINGW64 build artifact:
https://github.com/mmuetzel/MINGW-packages/actions/runs/2943110924

That doesn't seem to make a difference. The test program still crashes with a segmentation fault:
```
$ lldb ./a.out
(lldb) target create "./a.out"
Current executable set to 'C:\msys64\home\Markus\a.out' (x86_64).
(lldb) r
Process 18272 launched: 'C:\msys64\home\Markus\a.out' (x86_64)
Process 18272 stopped
* thread #1, stop reason = Exception 0xc0000005 encountered at address 0x7ff80a3cf5a3: Access violation reading location 0x1eb793cd010
    frame #0: 0x00007ff80a3cf5a3 libopenblas.dll`zgemv_n_ZEN + 5187
libopenblas.dll`zgemv_n_ZEN:
->  0x7ff80a3cf5a3 <+5187>: vinsertf128 $0x1, (%r9), %ymm0, %ymm0
    0x7ff80a3cf5a9 <+5193>: movq   0x88(%rsp), %rdi
    0x7ff80a3cf5b1 <+5201>: vpermpd $0x11, %ymm0, %ymm2       ; ymm2 = ymm0[1,0,1,0]
    0x7ff80a3cf5b7 <+5207>: vpermpd $0x44, %ymm0, %ymm0       ; ymm0 = ymm0[0,1,0,1]
(lldb) bt
error: a.out {0x000022af}: DIE has DW_AT_ranges(DW_FORM_sec_offset 0x1a) attribute, but range extraction failed (invalid range list offset 0x1a), please file a bug and attach the file at the start of this error message
error: a.out {0x000022af}: DIE has DW_AT_ranges(DW_FORM_sec_offset 0x1a) attribute, but range extraction failed (invalid range list offset 0x1a), please file a bug and attach the file at the start of this error message
* thread #1, stop reason = Exception 0xc0000005 encountered at address 0x7ff80a3cf5a3: Access violation reading location 0x1eb793cd010
  * frame #0: 0x00007ff80a3cf5a3 libopenblas.dll`zgemv_n_ZEN + 5187
    frame #1: 0x00007ff808d9bbff libopenblas.dll`zgemv_ + 1039
    frame #2: 0x00007ff80b1ebd67 libopenblas.dll`zlatrd_ + 2455
    frame #3: 0x00007ff80b154ebd libopenblas.dll`zhetrd_ + 2093
    frame #4: 0x00007ff80b149ddb libopenblas.dll`zheev_ + 859
    frame #5: 0x00007ff6a09418c5 a.out`main
    frame #6: 0x00007ff6a09413ae a.out`__tmainCRTStartup
    frame #7: 0x00007ff6a09414e6 a.out`mainCRTStartup
    frame #8: 0x00007ff8795254e0 kernel32.dll`BaseThreadInitThunk + 16
    frame #9: 0x00007ff87b46485b ntdll.dll`RtlUserThreadStart + 43
(lldb)
```

I wonder why it didn't crash when I rebuilt OpenBLAS locally. Is this a peculiarity of my system? Does rebuilding locally help for anyone else?

**Comment by mmuetzel** [comment]
Maybe https://github.com/xianyi/OpenBLAS/issues/3740?

**Comment by mmuetzel** [comment]
I let the CI rebuild OpenBLAS with less aggressive optimization (`-O2 -fno-tree-vectorize` as opposed to `-O3` for the release build):
https://github.com/mmuetzel/MINGW-packages/actions/runs/2944647955

After installing that MINGW64 build artifact, the test program no longer segfaults locally. 🎉

I don't know what exactly I did for the local build that worked for me. Rebuilding without stripping symbols no longer avoids the issue for me now. Unfortunately, I overwrote the package that was working for me. Maybe I built with debug settings (i.e., `-Og` afaict)?

Anyway, I'll try and clean up the PKGBUILD used for the tests and will submit a PR when ready.
